istioctl/pkg/waypoint/waypoint.go

```diff
@@ -49,33 +50,63 @@ var (
 
 	deleteAll bool
 
-	trafficType       string
+	trafficType       = constants.ServiceTraffic
 	validTrafficTypes = sets.New(constants.ServiceTraffic, constants.WorkloadTraffic, constants.AllTraffic, constants.NoTraffic)
 
-	waypointName string
+	waypointName      = constants.DefaultNamespaceWaypoint
+	enrolledNamespace = constants.DefaultEnrolledNamespace
 )
 
 const waitTimeout = 90 * time.Second
 
 func Cmd(ctx cli.Context) *cobra.Command {
-	makeGatewayName := func(name string) string {
-		if name == "" {
-			name = constants.DefaultNamespaceWaypoint
+	annotateNamespaceWithWaypoint := func(kubeClient kube.CLIClient) error {
+		nsObj, err := kubeClient.Kube().CoreV1().Namespaces().Get(context.Background(), enrolledNamespace, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("namespace: %s not found", enrolledNamespace)
+		} else if err != nil {
+			return fmt.Errorf("failed to get namespace %s: %v", enrolledNamespace, err)
 		}
-		return name
+		if nsObj.Annotations == nil {
+			nsObj.Annotations = map[string]string{}
+		} else if _, exists := nsObj.Annotations[constants.AmbientUseWaypoint]; exists && enrolledNamespace == constants.DefaultEnrolledNamespace {
+			// If the namespace is already annotated with a Waypoint name and the user did not specify a Waypoint name, return early.
+			return nil
+		}
+		nsObj.Annotations[constants.AmbientUseWaypoint] = waypointName
+		return nil
 	}
 	makeGateway := func(forApply bool) (*gateway.Gateway, error) {
 		ns := ctx.NamespaceOrDefault(ctx.Namespace())
 		if ctx.Namespace() == "" && !forApply {
 			ns = ""
 		}
+		// If a user sets the waypoint name to an empty string, set it to the default namespace waypoint name.
+		if waypointName == "" {
+			waypointName = constants.DefaultNamespaceWaypoint
+		}
+		// If a user sets the enrolled namespace to an empty string, set it to the default enrolled namespace.
+		if enrolledNamespace == "" {
+			enrolledNamespace = constants.DefaultEnrolledNamespace
+		}
+		kubeClient, err := ctx.CLIClient()
+		if err != nil {
+			return nil, fmt.Errorf("failed to create Kubernetes client: %v", err)
+		}
+		// If waypoint should be enrolled into specific namespace, annotate that namespace with the waypoint;
+		// otherwise, the default is to enroll the waypoint into the default namespace.
+		err = annotateNamespaceWithWaypoint(kubeClient)
+		if err != nil {
+			log.Errorf("failed to annotate namespace with waypoint: %v", err)
+		}
```

Comment:
I think we would want to annotate the namespace after we try applying the waypoint so that if the waypoint fails to apply we can avoid annotating the namespace. It's just less cleanup on an unhappy path.

Reply:
Is the suggestion to do it outside of the `makeGateway` function?

Reply:
Seems to me like the end of applyWaypointCmd would be a good spot if we like the idea of conditionally annotating only after the waypoint is successfully deployed.